lib/ExpressionUtils/core/exprMatrix_utils.py

```diff
@@ -203,3 +207,117 @@ def get_expression_matrix(self, params):
+
+    def convert_dem_to_dict( self, dem ):
+
+        row_ids = dem.get( 'row_ids' )
+        vals = dem.get( 'values' )
+  
+        n_rows = len( row_ids )
+        if ( len( vals ) != n_rows ):
+            raise Exception( "length discrepancy in differential expression matrix: {0} row_ids but {1} values".format( n_rows, len( fvals ) ) )
+
+        dem_dict = {}
+        for i in range( 0, n_rows ):
+            dem_dict[ row_ids[i] ] = [ vals[i][0], vals[i][2] ] # [fc,q]. (not bothering to check for dups here)
+
+        return dem_dict
+
+    # implments get_enhancedFilteredExpressionMatrix() method
+
+    def get_enhancedFEM( self, params ):
+
+        if not params.get( 'fem_object_ref' ):
+            raise ValueError( "fem_object_ref parameter not given to get_enhancedFilteredExpressionMatrix" )
+
+        fem_object_ref = params.get( 'fem_object_ref' )
+
+        fem_obj_ret = self.ws_client.get_objects2(
+                       {'objects': [{'ref': fem_object_ref }]})['data'][0]
+        fem = fem_obj_ret.get( 'data' )
+        prov = fem_obj_ret.get( 'provenance')[0]
+
+        # (1) create the enhanced FEM, starting with the FEM
+
+        efem = {}
+        for k in [ 'genome_ref', 'scale', 'type' ]:
+            efem[k] = fem.get( k )
+
+        efem['data'] = {}
+        efem['data']['col_ids'] = [ "description", 
+                                    "fold-change",
+                                    "q-value",
+                                    "min",
+                                    "max",
+                                    "mean",
+                                    "std_dev",
+                                    "is_missing_values" ]
+        efem['data']['column_labels'] =[ "Description", 
+                                         "Fold change",
+                                         "Q value",
+                                         "Min. expression",
+                                         "Max. expression",
+                                         "Mean expression",
+                                         "Std. dev.",
+                                         "Missing values?" ]
+        fm = fem.get('data')
+        efem['data']['row_ids'] = fm.get('row_ids')
+        efem['data']['values' ] = []
+        n_efem_rows = len( efem['data']['row_ids'] )
+        fvals = fm.get('values')
+        if ( len( fvals ) != n_efem_rows ):
+            raise Exception( "length discrepancy in filtered expression matrix: {0} row_ids but {1} values".format( n_efem_rows, len( fvals ) ) )
+
+        for i in range( 0, n_efem_rows ):
+            efem['data']['values'].append( [ 'NA', 'NA', 'NA' ] + self.get_matrix_stats( fvals[i] ) )
+
+        # (2) Get genome object and feature descriptions, put those in column 1
+
+        feat_dict = self.gaa.get_feature_functions( { 'ref': fem.get( 'genome_ref' ), 'feature_id_list': None } )
+        for i in range( 0, n_efem_rows ):
```

Comment:
It's more python to iterate the items than the index. You can use zip here again.